### DISCUSSION TITLE
ng build not working to compile vitrivr-ng from the source

### DISCUSSION THREAD

**Discussion by newtonsp** [ppanopticon/vitrivr-ng#79] [Q&A] [answered]
Dear developers,

Sorry by disturbing you again. I'm trying to build vitrivr-ng from the source without success. These are the steps I made:
- I followed the "Getting started" section indicated in [https://github.com/vitrivr/vitrivr-ng/wiki](url) . After the last one, I found 35 moderate severity vulnerabilities (if needed, you can check the log file "npm_install.txt" and "npm_install_error.txt" created with the command npm install > npm_install.txt 2> npm_install_error.txt ).
[npm_install.txt](https://github.com/vitrivr/vitrivr-ng/files/6550022/npm_install.txt)
[npm_install_error.txt](https://github.com/vitrivr/vitrivr-ng/files/6550035/npm_install_error.txt)

- To better understand the vulnerabilities, I got one more non-empty log file (npm_audit.txt) by running the following command npm audit > npm_audit.txt 2> npm_audit_error.txt
[npm_audit.txt](https://github.com/vitrivr/vitrivr-ng/files/6550044/npm_audit.txt)

- If I run ng build in this scenario, the "ng_build_error.txt" non-empty file is generated with the command ng build > ng_build.txt 2> ng_build_error.txt
[ng_build_error.txt](https://github.com/vitrivr/vitrivr-ng/files/6550070/ng_build_error.txt)

I spent some time trying to manually fix the dependencies and running npm update, but I couldn't solve the issue. 

My environment is:
- Windows 10 (win32) x64
- NPM 6.14.8
- Node 14.16.1
- Angular CLI 11.0.6
- Angular 11.2.14

Do you have any suggestion to solve the problem? If you need more information, I can share it.

Thank you so much.
Sincerely

**Comment by lucaro** [comment, marked as answer]
If I understand your issue correctly, the build is actually working fine and produces the compiled project but generates some warnings about potential security issues. This is not unexpected since, especially on the main branch, not all dependencies are used in their most recent version and the older might contain known issues. This is not a priority for us, since vitrivr instances are rarely deployed in a public, outward facing way and also do not deal with any data where such security issues could be meaningfully exploited. If you want to resolve these issues, a first step to try would be to see if any dependencies can be updated. I hope this clarifies the matter. Please correct me in case I misunderstood your issue.

**Reply by newtonsp** [reply]
Dear Luca,

Thank you for your answer.

Maybe the problem is simpler than I thought then, if the warnings can be ignored. But I think the project is not compiling, as a "dist" folder was not generated in the end. Let me show you the current project folder in an attached image. 
![print](https://user-images.githubusercontent.com/20126691/119815667-2d6d4200-bec2-11eb-9a6a-5d75c5a1348d.png)

Maybe it's a configuration issue... If I understood the angular.json file (attached as angular.txt to pass the GitHub filter), dist should be the output directory. 
[angular.txt](https://github.com/vitrivr/vitrivr-ng/files/6553453/angular.txt)

Thank you so much.
Sincerely

**Reply by lucaro** [reply]
Did you run `ng build` with the `prod` flag? Otherwise it won't produce the 'dist' folder.

**Reply by newtonsp** [reply]
Dear Luca,

Thanks. In the past I did, and I just did it again (ng build -- prod). Unfortunately, the terminal returned the same messages I attached before (ng_build_error.txt) and the dist folder was not created.

I tried a few variations in the past, like creating an empty dist folder, and nothing changed. The last attempt started with the same directory setting shown in the previous figure.

If I didn't do a configuration error, maybe some dependencies should be solved. The problem is that I couldn't do it. Automatic approaches didn't work. After a few hours, the manual one didn't work either.

sincerely

**Reply by lucaro** [reply]
Looking through your error log again, none of these errors appear to originate from vitrivr-ng but rather from some dependencies. It looks like something must have gone wrong when installing those. Have you tried clearing the npm dependency cache and re-installing them?

**Reply by newtonsp** [reply]
Dear Luca,

Thank you so much.

I tried two approaches based on your suggestion. Please correct me if I didn't understand it well.

1)First, I tried to clean the cache after verifying it with the following commands:
npm cache verify

The output was:
Cache verified and compressed (~\AppData\Roaming\npm-cache\_cacache):
Content verified: 3009 (284254315 bytes)
Corrupted content removed: 4
Content garbage-collected: 262 (99415119 bytes)
Missing content: 4
Index entries: 4723
Finished in 160.421s

Then I ran additional commands:
npm cache clean --force
ng build --prod

The last command led to the same messages previously attached

2)Second, I tried to remove node_modules directory and reinstall the modules. I already manually removed the node_modules in the past, but I tried now using the following command in vitrivr-ng folder:
rm -rf node_modules && npm install

A few warning/errors happened, as indicated in what follows:
Warning: Unable to fully load E:/Programming/Posdoc/vitrivr-ng-vs/node_modules/@angular-slider/ngx-slider/bundles/angular-slider-ngx-slider.umd.js for source-map flattening: Cannot read property 'replace' of null
Warning: Unable to fully load E:/Programming/Posdoc/vitrivr-ng-vs/node_modules/@angular-slider/ngx-slider/bundles/angular-slider-ngx-slider.umd.js for source-map flattening: Cannot read property 'replace' of null
Error: Error when flattening the source-map "E:/Programming/Posdoc/vitrivr-ng-vs/node_modules/@angular-slider/ngx-slider/bundles/angular-slider-ngx-slider.umd.js.map" for "E:/Programming/Posdoc/vitrivr-ng-vs/node_modules/@angular-slider/ngx-slider/bundles/angular-slider-ngx-slider.umd.js": TypeError: Cannot read property 'rawMap' of null
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.13 (node_modules\webpack-dev-server\node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.13: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.3.2 (node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})

I noted that 35 moderate severity vulnerabilities were identified again. Then I ran the following command:
ng build --prod > ng_build_prod.txt 2> ng_build_prod_err.txt

Unfortunately, the last command led to messages similar to the ones previously attached. I'm attaching the last non-empty file, ng_build_prod_err.txt
[ng_build_prod_err.txt](https://github.com/vitrivr/vitrivr-ng/files/6555983/ng_build_prod_err.txt)

I'd like to add that the first step in Getting started (git clone) was made from Visual Studio 2019. But I think this is not problematic, as before the issues already indicated in this discussion, I was using only CMD. Even by doing so, ng build and ng build --prod were not compiling the project.

I agree that the problem can be in the dependencies. But do you think I missed a vitrivr-ng requirement (e.g. a specific Angular version) in my environment (1st message in this discussion)? Depending on the answer, I could for example reinstall npm, Angular and related programs with a specific version.

Sincerely

**Reply by silvanheller** [reply]
Hey Newton, Unfortunately I cannot reproduce your issue. Looking at your error log, it seems to me that `npm install` has failed since the error log mostly complains about dependencies not being installed. We've had multiple student projects in the past months that have not encountered this issue. Beyond reinstalling npm, I do not have a solution to the issue. Can your reproduce the issue on another machine you have?

**Reply by newtonsp** [reply]
Dear Silvan and Luca,

Sorry the delay, andthank you for your suggestions. It works now! I built vitrivr-ng with success after reinstalling Node.js and NPM in the same machine from the first message.

The current versions are:
Angular CLI: 12.0.2
Node: 14.17.0
npm: 6.14.8

Best regards
